### PR TITLE
Allow more generic searches of ACLs

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #8859 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by kiranchavala — COMMENTED** [review]
@DaanHoogland  currently the keyword search is working only for protocol field and description 

aclsearch


can the keyword search improved to other parameters like 

```
(cmk) > list networkacls -h
listNetworkACLs: Lists all network ACL items
API Params               Type     Description
==========               ====     ===========
account                  string   list resources by account. Must be used w
                                  ith the domainId parameter.
aclid                    uuid     list network ACL items by ACL ID
action                   string   list network ACL items by action
domainid                 uuid     list only resources belonging to the doma
                                  in specified
fordisplay               boolean  list resources by display flag; only ROOT
                                   admin is eligible to pass this paramete
                                  r
id                       uuid     Lists network ACL Item with the specified
                                   ID
isrecursive              boolean  defaults to false, but if true, lists all
                                   resources from the parent specified by
                                  the domainId till leaves.
keyword                  string   List by keyword
listall                  boolean  If set to false, list only resources belo
                                  nging to the command's caller; if set to
                                   true - list resources that the caller i
                                  s authorized to see. Default value is fa
                                  lse. Resources dedicated to a project ar
                                  e listed only if using the projectid par
                                  ameter.
networkid                uuid     list network ACL items by network ID
page                     integer
pagesize                 integer
projectid                uuid     list objects by project; if projectid=-1
                                  lists All VMs
protocol                 string   list network ACL items by protocol
tags                     map      List resources by tags (key/value pairs)
traffictype              string   list network ACL items by traffic type -
                                  ingress or egress
```

**Comment by DaanHoogland** [comment]
@kiranchavala that would be a bigger effort, the search is no only for acl item fields. As for account, aclid, domain etc, those are implied (at least in the GUI). things like traffic_type and action, these are enumerals and require a different kind of selection, i.e. they are not free text searchable and hence are left out of scope.

**Review by bernardodemarco — COMMENTED** [review]
I manually tested in a local environment.

Regarding the `listNetworkACLs` API, the filtering is working as expected. I created four ACL rules and filtered them by protocol and description.

All ACL rules:
image

Filtering by description:
image

Filtering by protocol:
image

@DaanHoogland, what do you think about using one of the Ant Design components for the input field?

---

Regarding the `listNetworkACLLists`, when executing the API with the `keyword` parameter, apparently a NPE is not being handled.

```bash
(localcloud) 🐱 > list networkacllists vpcid="f9510528-eb2c-405f-b690-21a540c00661" keyword="acl"
🙈 Error: (HTTP 530, error code 4250) Caught error: null
```

Management server logs:
```bash
2024-09-01T16:09:42,439 ERROR [c.c.n.v.d.NetworkACLDaoImpl] (qtp1438988851-385:[ctx-49b0b1cf, ctx-54f34d08]) (logid:d0c03e10) Caught: com.mysql.cj.jdbc.ClientPreparedStatement: SELECT network_acl.id, network_acl.vpc_id, network_acl.uuid, network_acl.name, network_acl.description, network_acl.display FROM network_acl WHERE network_acl.vpc_id IN (1,0)  AND network_acl.display = 1  AND  (network_acl.name LIKE x'2561636c25'  OR network_acl.description LIKE x'2561636c25'  OR )  ORDER BY network_acl.id DESC  LIMIT 0, 500 java.lang.NullPointerException

2024-09-01T16:09:42,439 ERROR [c.c.a.ApiServer] (qtp1438988851-385:[ctx-49b0b1cf, ctx-54f34d08]) (logid:d0c03e10) unhandled exception executing api command: [Ljava.lang.String;@33af317 com.cloud.utils.exception.CloudRuntimeException: Caught error: null

```

**Comment by DaanHoogland** [comment]
> I manually tested in a local environment.

thanks @bernardodemarco 



> @DaanHoogland, what do you think about using one of the Ant Design components for the input field?

which one would you like to see implemented?

> Regarding the `listNetworkACLLists`, when executing the API with the `keyword` parameter, apparently a NPE is not being handled.
> 
> ```shell
> (localcloud) 🐱 > list networkacllists vpcid="f9510528-eb2c-405f-b690-21a540c00661" keyword="acl"
> 🙈 Error: (HTTP 530, error code 4250) Caught error: null
> ```
> 
> Management server logs:
> 
```shell
2024-09-01T16:09:42,439 ERROR [c.c.n.v.d.NetworkACLDaoImpl] (qtp1438988851-385:[ctx-49b0b1cf, ctx-54f34d08]) (logid:d0c03e10) Caught: com.mysql.cj.jdbc.ClientPreparedStatement: SELECT network_acl.id, network_acl.vpc_id, network_acl.uuid, network_acl.name, network_acl.description, network_acl.display FROM network_acl WHERE network_acl.vpc_id IN (1,0)  AND network_acl.display = 1  AND  (network_acl.name LIKE x'2561636c25'  OR network_acl.description LIKE x'2561636c25'  OR )  ORDER BY network_acl.id DESC  LIMIT 0, 500 java.lang.NullPointerException
2024-09-01T16:09:42,439 ERROR [c.c.a.ApiServer] (qtp1438988851-385:[ctx-49b0b1cf, ctx-54f34d08]) (logid:d0c03e10) unhandled exception executing api command: [Ljava.lang.String;@33af317 com.cloud.utils.exception.CloudRuntimeException: Caught error: null
```

thanks, will have a look

**Comment by bernardodemarco** [comment]
> which one would you like to see implemented?

I believe that the `a-input` or `a-input-search` components could be used, what do you think?

**Comment by DaanHoogland** [comment]
@bernardodemarco took some time (priorities) but updated the PR now.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@kiranchavala @bernardodemarco is this ok by you? (cc @JoaoJandre )

**Review by bernardodemarco — APPROVED** [review]
@DaanHoogland, I've tested it again, and everything seems to be working well.

All ACL rules:
image

---

Filtering by protocol:
image

---

Filtering by description:
image

---

Execution of the `listNetworkACLLists` API with the `keyword` parameter:
```bash
(localcloud) 🐱 > list networkacllists keyword="vpc"
{
  "count": 1,
  "networkacllist": [
    {
      "description": "ACL 01",
      "fordisplay": true,
      "id": "e85bc061-1ffe-475a-92f0-a7dfbcfbd878",
      "name": "vpc-01-acl-01"
    }
  ]
}
```

**Comment by kiranchavala** [comment]
> @kiranchavala that would be a bigger effort, the search is no only for acl item fields. As for account, aclid, domain etc, those are implied (at least in the GUI). things like traffic_type and action, these are enumerals and require a different kind of selection, i.e. they are not free text searchable and hence are left out of scope.



Thanks @DaanHoogland  we can revisit them in future if required

File: server/src/main/java/com/cloud/network/vpc/NetworkACLServiceImpl.java [hunk 1]
```diff
@@ -730,24 +732,31 @@ public Pair<List<? extends NetworkACLItem>, Integer> listNetworkACLItems(final L
...
 
-        if (id != null) {
-            sc.setParameters("id", id);
-        }
+        if (StringUtils.isNotBlank(keyword)) {
+            final SearchCriteria<NetworkACLItemVO> ssc = _networkACLItemDao.createSearchCriteria();
+            ssc.addOr("protocol", SearchCriteria.Op.LIKE, "%" + keyword + "%");
+            ssc.addOr("reason", SearchCriteria.Op.LIKE, "%" + keyword + "%");
+            sc.addAnd("acl_id", SearchCriteria.Op.SC, ssc);
+        } else {
```
Comment (vishesh92) [inline server/src/main/java/com/cloud/network/vpc/NetworkACLServiceImpl.java:740]: @DaanHoogland do we need an else here?

**Comment by vishesh92** [comment]
@blueorangutan ui

File: ui/src/views/network/AclListRulesTab.vue [hunk 2]
```diff
@@ -31,6 +32,14 @@
         <template #icon><download-outlined /></template>
         {{ $t('label.acl.export') }}
       </a-button>
+      <div class="search-bar">
+        <a-input-search
+          style="width: 25vw;float: right;margin-bottom: 10px; z-index: 8"
```
Comment (vishesh92) [inline ui/src/views/network/AclListRulesTab.vue:37]:
```suggestion
          style="width: 25vw;float: right;margin-left: 10px; z-index: 8"
```

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by vishesh92 — APPROVED** [review]
lgtm. Tested on my local env with simulator. Search with keyword is working as expected.

**Pull request merged**